server/s3_data_usage.cc

```diff
@@ -0,0 +1,511 @@
+  if (current_objects_increment == 0 && current_bytes_increment == 0) {
+    run_successful_callbacks();
+    current_request = nullptr;
+    set_state(DataUsageItemState::inactive);
+    s3_log(S3_LOG_DEBUG, req_id, "%s Exit\n", __func__);
+    return;
+  }
+  set_state(DataUsageItemState::active);
+  S3MotrKVSWriterFactory writer_factory;
+  motr_kv_writer =
+      writer_factory.create_motr_kvs_writer(current_request, nullptr);
+  motr_kv_writer->put_keyval(
+      data_usage_accounts_index_layout, motr_key, this->to_json(),
+      std::bind(&DataUsageItem::kvs_write_success, this),
+      std::bind(&DataUsageItem::kvs_write_failure, this));
+
+  s3_log(S3_LOG_DEBUG, req_id, "%s Exit\n", __func__);
+}
+
+void DataUsageItem::kvs_write_success() {
+  std::string req_id = get_item_request_id();
+  s3_log(S3_LOG_DEBUG, req_id, "%s Entry\n", __func__);
+
+  // This is so that we dont have to call
+  // getkeyval everytime. We only do it once
+  // then we maintain things in memory.
+  // this may be time efficient.
+  run_successful_callbacks();
+  objects_count += current_objects_increment;
+  bytes_count += current_bytes_increment;
+  current_objects_increment = 0;
+  current_bytes_increment = 0;
+  current_request = pending_request;
+  pending_request = nullptr;
+  // Call unconditionally. do_kvs_write can handle empty pending queue.
+  do_kvs_write();
+
+  s3_log(S3_LOG_DEBUG, req_id, "%s Exit\n", __func__);
+}
+
+void DataUsageItem::kvs_write_failure() {
+  std::string req_id = get_item_request_id();
+  s3_log(S3_LOG_DEBUG, req_id, "%s Entry\n", __func__);
+  run_failure_callbacks();
+  current_objects_increment = 0;
+  current_bytes_increment = 0;
+  current_request = pending_request;
+  pending_request = nullptr;
+  // Try to proceed other requests.
+  do_kvs_write();
+  s3_log(S3_LOG_DEBUG, req_id, "%s Exit\n", __func__);
+}
+
+std::string DataUsageItem::to_json() {
+  std::string req_id = get_item_request_id();
+  s3_log(S3_LOG_DEBUG, req_id, "%s Entry\n", __func__);
+
+  Json::Value root;
+  root[JSON_OBJECTS_COUNT] = Json::Value(
+      (Json::Value::UInt64)(objects_count + current_objects_increment));
+  root[JSON_BYTES_COUNT] =
+      Json::Value((Json::Value::UInt64)(bytes_count + current_bytes_increment));
+
+  Json::FastWriter fastWriter;
+  s3_log(S3_LOG_DEBUG, req_id, "[%s] Exit\n", __func__);
+  return fastWriter.write(root);
+}
+
+int DataUsageItem::from_json(std::string content) {
+  std::string req_id = get_item_request_id();
+  s3_log(S3_LOG_DEBUG, req_id, "%s Entry\n", __func__);
+  s3_log(S3_LOG_DEBUG, req_id, "Called with content [%s]\n", content.c_str());
+  Json::Value newroot;
+  Json::Reader reader;
+  bool parsingSuccessful = reader.parse(content.c_str(), newroot);
+  if (!parsingSuccessful) {
+    s3_log(S3_LOG_ERROR, req_id, "Json Parsing failed\n");
+    return -1;
+  }
+
+  objects_count = newroot[JSON_OBJECTS_COUNT].asUInt64();
+  bytes_count = newroot[JSON_BYTES_COUNT].asUInt64();
```

Comment:
Add validation that these fields are actually present in the JSON.

Reply:
Done.